### PR TITLE
ScalarEncoder handling of n/radius/resolution when maxval/minval is None

### LINKED ISSUE #1966: Inconsistent ScalarEncoder constructor behavior
Buried in `ScalarEncoder._initEncoder()`, there is the possibility for raising a `ValueError` exception suggesting that one of `n`, `radius`, or `resolution` is required, but it seems `n` is always required.  For example, if you pass `n=0`, and specify either `radius` or `resolution` to something non-zero, it raises an exception: `AttributeError: 'ScalarEncoder' object has no attribute 'rangeInternal'`

This passes, resulting in a successfully created `ScalarEncoder` instance:

```python
encoder = ScalarEncoder(3, None, None, name='scalar', n=14, radius=0, resolution=0, forced=True)
```

These fail, raising unexpected `AttributeError` exceptions:

```python
encoder1 = ScalarEncoder(3, None, None, name='scalar', n=0, radius=None, resolution=0.5, forced=True)
encoder2 = ScalarEncoder(3, None, None, name='scalar', n=0, radius=1.5, resolution=0, forced=True)
```

### PR DESCRIPTION
**Pull request by rmsmith007**
Fixes #1966

### CODE REVIEW

File: nupic/encoders/scalar.py [hunk 1]
```diff
@@ -253,8 +254,8 @@ def _initEncoder(self, w, minval, maxval, n, radius, resolution):
     """ (helper function)  There are three different ways of thinking about the representation.
      Handle each case here."""
     if n != 0:
-      assert radius == 0
-      assert resolution == 0
+      if (radius !=0 or resolution != 0):
+        raise Exception("ScalarEncoder: n is specified, so radius and resolution should be default")
```
Comment (oxtopus) [inline nupic/encoders/scalar.py]: This, and other exceptions related to checking the value should be [ValueError](https://docs.python.org/2/library/exceptions.html#exceptions.ValueError) exceptions.

File: tests/unit/nupic/encoders/scalar_test.py [hunk 2]
```diff
@@ -448,6 +448,25 @@ def testReadWrite(self):
...
+  # Tests for #1966
+  def testSettingNWithMaxvalMinvalNone(self):
+    """Setting n when maxval/minval = None creates instance."""
+    encoder = ScalarEncoder(3, None, None, name='scalar',
+                            n=14, radius=0, resolution=0, forced=True)
+    self.assertIsInstance(encoder, ScalarEncoder)
+
+  def testSettingScalarAndResolution(self):
+    """Setting both scalar and resolution not allowed."""
+    with self.assertRaises(Exception):
```
Comment (oxtopus) [inline tests/unit/nupic/encoders/scalar_test.py]: Assert that it raises `ValueError`.

**Comment by oxtopus** [comment]
This looks good to me, but please raise `ValueError` exceptions instead of the generic `Exception`.

Thanks!

**Comment by rhyolight** [comment]
Updated description with proper "fixes" nomenclature.

**Comment by rmsmith007** [comment]
@oxtopus @rhyolight OK, 'tis changed, so I am open to take on something else. I can devote about 10 hours per week to nupic, but to start with I'd prefer easily digestible tasks. I guess I could just scan under the help wanted tag, but then there would be no element of surprise.

**Comment by rhyolight** [comment]
@rmsmith007 Congrats on your first commit to NuPIC!

**Comment by rmsmith007** [comment]
@rhyolight Thanks, I shall find an appropriate way to celebrate. It's actually my first community commit in almost 5 years of github membership. I think I started using it for Heroku and later for Phonegap.

**Pull request merged**